From 0d9d5aef08d8171e1a32958514925cce933a5aca Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Tue, 6 Oct 2020 23:16:43 +0800
Subject: [PATCH] DRVLIB_SW-3833:XPCS: test pattern and prbs support

Defined test pattern and Pseudo Random Binary Sequence(PRBS)
test is standard feature to verify link quality, this feature enables
support in XPCS driver

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/ethernet/lantiq/xpcs/xpcs.h       |  38 ++++
 drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c | 268 ++++++++++++++++++++++++++
 2 files changed, 306 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.h b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
index 25ef0f6670dd..68e7610459a2 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.h
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
@@ -575,6 +575,44 @@
 #define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_POS	12
 #define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_WIDTH	4
 
+/* Test Pattern */
+#define PCS_TP_CTRL_TP_SEL_POS	1
+#define PCS_TP_CTRL_TP_SEL_WIDTH	1
+#define PCS_TP_CTRL_RTP_EN_POS	2
+#define PCS_TP_CTRL_RTP_EN_WIDTH	1
+#define PCS_TP_CTRL_TTP_EN_POS	3
+#define PCS_TP_CTRL_TTP_EN_WIDTH	1
+#define PCS_TP_CTRL_PRBS31T_EN_POS	4
+#define PCS_TP_CTRL_PRBS31T_EN_WIDTH	1
+#define PCS_TP_CTRL_PRBS31R_EN_POS	5
+#define PCS_TP_CTRL_PRBS31R_EN_WIDTH	1
+#define PCS_TP_CTRL_PRBS9T_EN_POS	6
+#define PCS_TP_CTRL_PRBS9T_EN_WIDTH	1
+#define PCS_KR_CTRL_PRBS9RXEN_POS	7
+#define PCS_KR_CTRL_PRBS9RXEN_WIDTH	7
+
+/* Tx Tx to Rx Loopback */
+#define PMA_CTRL1_LB_POS	0
+#define PMA_CTRL1_LB_WIDTH	1
+/* Rx to Tx Loopback */
+#define PCS_DIG_CTRL1_R2TLBE_POS	14
+#define PCS_DIG_CTRL1_R2TLBE_WIDTH	1
+/* Square wave period */
+#define PCS_KR_CTRL_NVAL_SEL_POS	4
+#define PCS_KR_CTRL_NVAL_SEL_WIDTH	3
+/* PCS Error Block Counter */
+#define PCS_KR_STS2_ERR_BLK_POS 0
+#define PCS_KR_STS2_ERR_BLK_WIDTH	8
+/*PCS Bit Error Rate Counter */
+#define PCS_KR_STS2_BER_CNT_POS	8
+#define PCS_KR_STS2_BER_CNT_WIDTH	6
+/* PCS Latched High Bit Error */
+#define PCS_KR_STS2_LAT_HBER_POS 14
+#define PCS_KR_STS2_LAT_HBER_WIDTH	1
+/* Latched Block Lock */
+#define PCS_KR_STS2_LAT_BL_POS 15
+#define PCS_KR_STS2_LAT_BL_WIDTH	1
+
 enum {
 	BACKPL_ETH_PCS = 0,
 	OTHER_PCS_CONFIG = 1
diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
index 6658ac5d8e5f..7407d6511a41 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
@@ -1125,6 +1125,273 @@ static ssize_t cl37_store(struct device * dev, struct device_attribute * attr,
 }
 static DEVICE_ATTR_RW(cl37);
 
+static void serdes_tx_test_pattern_cfg_get(struct xpcs_prv_data *pdata)
+{
+	int val;
+
+	val = XPCS_RGRD(pdata, PCS_TP_CTRL) & 0xFFFF;
+	if (val & 0xA)
+		pr_info("\tTX bert mode = BERT_SQUAREWAVE_TX");
+	else if (val & 0x10)
+		pr_info("\tTX bert mode = BERT_PRBS31_TX");
+	else if (val & 0x40)
+		pr_info("\tTX bert mode = BERT_PRBS9_TX");
+	else
+		pr_info("\tTX bert mode = BERT_TX_MODE_DISABLE");
+}
+
+static void serdes_rx_test_pattern_cfg_get(struct xpcs_prv_data *pdata)
+{
+	int val;
+
+	val = XPCS_RGRD(pdata, PCS_TP_CTRL) & 0xFFFF;
+	if (val & 0x20)
+		pr_info("\tTX bert mode = BERT_PRBS31_RX");
+	val = XPCS_RGRD(pdata, PCS_KR_CTRL) & 0xFFFF;
+	if (val & 0x80)
+		pr_info("\tTX bert mode = BERT_PRBS9_TX");
+}
+
+static void serdes_rx_bit_error_get(struct xpcs_prv_data *pdata)
+{
+	int val;
+
+	val = XPCS_RGRD(pdata, PCS_TP_ERRCR);
+	pr_info("tp_error_cnt	= 0x%x\n", val);
+	val = XPCS_RGRD(pdata, PCS_KR_STS2);
+	val = XPCS_RGRD_VAL(pdata, PCS_KR_STS2, ERR_BLK);
+	pr_info("error_block_count	= 0x%x\n", val);
+	val = XPCS_RGRD_VAL(pdata, PCS_KR_STS2, BER_CNT);
+	pr_info("bit_error_rate_counter	=0x%x\n", val);
+	val = XPCS_RGRD_VAL(pdata, PCS_KR_STS2, LAT_HBER);
+	pr_info("latched_high_bit_err_rate	=0x%x\n", val);
+	val = XPCS_RGRD_VAL(pdata, PCS_KR_STS2, LAT_BL);
+	pr_info("latched_block_lock	=0x%x\n", val);
+}
+
+static ssize_t bert_show(struct device *dev, struct device_attribute *attr,
+			 char *buf)
+{
+	ssize_t off = 0;
+	const char *str;
+	int val;
+	struct xpcs_prv_data *pdata = dev_get_drvdata(dev);
+
+	if (!pdata->state)
+		return sprintf(buf, "\n%s: power off\n", pdata->name);
+
+	str = XPCS_RGRD_VAL(pdata, PCS_TP_CTRL, TP_SEL) ?
+		"enabled" : "disabled";
+	off += sprintf(buf + off, "\tTx_test_pattern:	%s\n",
+							  str);
+	str = XPCS_RGRD_VAL(pdata, PCS_TP_CTRL, RTP_EN) ?
+		"enabled" : "disabled";
+	off += sprintf(buf + off, "\tRx_test_pattern:	%s\n",
+							  str);
+	str = XPCS_RGRD_VAL(pdata, PCS_DIG_CTRL1, R2TLBE) ? "enabled" : "disabled";
+	off += sprintf(buf + off, "\tRemote loopback:	%s\n",
+							  str);
+	str = XPCS_RGRD_VAL(pdata, PMA_CTRL1, LB) ?
+		"enabled" : "disabled";
+	off += sprintf(buf + off, "\tLocal loopback:	%s\n",
+							  str);
+	val = XPCS_RGRD(pdata, PCS_TP_CTRL) & 0xFFFF;
+	if (val & 0xA)
+		off += sprintf(buf + off, "\tTX BERT MODE: %s\n",
+								  "\tBERT_SQUAREWAVE_TX");
+	else if (val & 0x10)
+		off += sprintf(buf + off, "\tTX BERT MODE: %s\n",
+								  "\tBERT_PRBS31_TX");
+	else if (val & 0x40)
+		off += sprintf(buf + off, "\tTX BERT MODE: %s\n",
+								  "\tBERT_PRBS39_TX");
+	else
+		off += sprintf(buf + off, "\tTX BERT MODE: %s\n",
+								  "\tBERT_TX_MODE_DISABLE");
+	val = XPCS_RGRD(pdata, PCS_TP_CTRL) & 0xFFFF;
+	if (val & 0x20)
+		off += sprintf(buf + off, "\tRX BERT MODE: %s\n",
+								  "\tBERT_PRBS31_RX");
+	val = XPCS_RGRD(pdata, PCS_KR_CTRL) & 0xFFFF;
+	if (val & 0x80)
+		off += sprintf(buf + off, "\tRX BERT MODE: %s\n",
+								  "\tBERT_PRBS9_RX");
+	return off;
+}
+
+static void bert_help(void)
+{
+	pr_info("Usage:\n");
+	pr_info("  echo <cmd>  > bert\n");
+	pr_info("	 cmd:\n");
+	pr_info("       tx_test_pattern_enable:		Enable Tx pattern\n");
+	pr_info("       tx_test_pattern_disable:	Disable Tx test pattern\n");
+	pr_info("       rx_test_pattern_enable:		Enable Rx test pattern\n");
+	pr_info("       rx_test_pattern_disable:	Disable Rx test pattern\n");
+	pr_info("       remote_loop_enable:			Enable parallel Rx to Tx loop\n");
+	pr_info("       remote_loop_disable:		Disable parallel Rx to Tx loop\n");
+	pr_info("       local_loop_enable:			Enable serial Rx to Tx loop\n");
+	pr_info("       local_loop_disable:			Disable serial Rx to Tx loop\n");
+	pr_info("       tx_test_pattern_cfg_set:	Define the pattern to be sent (PRBS/fixed pattern)\n");
+	pr_info("       tx_test_pattern_cfg_get:	Get Tx BERT mode\n");
+	pr_info("       rx_test_pattern_cfg_set:	Define the pattern to be received (PRBS/fixed pattern)\n");
+	pr_info("       rx_test_pattern_cfg_get:	Get Rx BERT mode\n");
+	pr_info("       rx_bit_error_counter_get:	Read the bit error counter,clear-on-read\n");
+}
+
+struct bert_cmd {
+	const char *name;
+	int val;
+};
+
+static struct bert_cmd bert_cmd_tbl[] = {
+	{"tx_test_pattern_enable", 0 },
+	{"tx_test_pattern_disable", 1},
+	{"rx_test_pattern_enable", 2},
+	{"rx_test_pattern_disable", 3},
+	{"remote_loop_enable", 4},
+	{"remote_loop_disable", 5},
+	{"local_loop_enable", 6},
+	{"local_loop_disable", 7},
+	{"tx_test_pattern_cfg_set", 8},
+	{"tx_test_pattern_cfg_get", 9},
+	{"rx_test_pattern_cfg_set", 10},
+	{"rx_test_pattern_cfg_get", 11},
+	{"rx_bit_error_counter_get", 12},
+};
+
+static int bert_cmd_idx_get(const char *name)
+{
+	struct bert_cmd *p;
+
+	for (p = bert_cmd_tbl; p->name; p++) {
+		if (strncasecmp(p->name, name, strlen(name)) == 0)
+			return p->val;
+	}
+
+	return -EINVAL;
+}
+
+static ssize_t bert_store(struct device *dev, struct device_attribute *attr,
+			  const char *buf, size_t count)
+{
+	struct xpcs_prv_data *pdata = dev_get_drvdata(dev);
+	char lbuf[256];
+	size_t len;
+	char *p, *token;
+	const char *ct = " \t\n";
+	int index;
+
+	len = min(count, sizeof(lbuf) - 1);
+	memcpy(lbuf, buf, len);
+	lbuf[len] = '\0';
+
+	p = lbuf;
+	token = strsep(&p, ct);
+	if (!token) {
+		bert_help();
+		return count;
+	}
+	len = strlen(token);
+	index = bert_cmd_idx_get(token);
+	switch (index) {
+	case 0:
+		XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TP_SEL, 1);
+		break;
+	case 1:
+		XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TP_SEL, 0);
+		break;
+	case 2:
+		XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, RTP_EN, 1);
+		break;
+	case 3:
+		XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, RTP_EN, 0);
+		break;
+	case 4:
+		XPCS_RGWR_VAL(pdata, PCS_DIG_CTRL1, R2TLBE, 1);
+		break;
+	case 5:
+		XPCS_RGWR_VAL(pdata, PCS_DIG_CTRL1, R2TLBE, 0);
+		break;
+	case 6:
+		XPCS_RGWR_VAL(pdata, PMA_CTRL1, LB, 1);
+		break;
+	case 7:
+		XPCS_RGWR_VAL(pdata, PMA_CTRL1, LB, 0);
+		break;
+	case 8:
+		for (token = strsep(&p, ct); token; token = strsep(&p, ct)) {
+			len = strlen(token);
+			if (!len)
+				continue;
+			if (strncasecmp(token, "square",
+					min_t(size_t, len, 6UL)) == 0) {
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TTP_EN, 1);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL,
+					      PRBS31T_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, PRBS9T_EN, 0);
+			} else if (strncasecmp(token, "PRBS31",
+				min_t(size_t, len, 6UL)) == 0) {
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TTP_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TP_SEL, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, PRBS9T_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL,
+					      PRBS31T_EN, 1);
+			} else if (strncasecmp(token, "PRBS9",
+				min_t(size_t, len, 5UL)) == 0) {
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TTP_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, TP_SEL, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL,
+					      PRBS31T_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, PRBS9T_EN, 1);
+			} else if (strncasecmp(token, "RANDOM",
+				min_t(size_t, len, 6UL)) == 0)
+				pr_info("WARN: random mode not supported");
+			}
+		break;
+	case 9:
+		serdes_tx_test_pattern_cfg_get(pdata);
+		break;
+	case 10:
+		for (token = strsep(&p, ct); token; token = strsep(&p, ct)) {
+			len = strlen(token);
+			if (!len)
+				continue;
+			if (strncasecmp(token, "PRBS31",
+					min_t(size_t, len, 6UL)) == 0) {
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, RTP_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_KR_CTRL, PRBS9RXEN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL,
+					      PRBS31R_EN, 1);
+
+			} else if (strncasecmp(token, "PRBS9",
+				min_t(size_t, len, 5UL)) == 0) {
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL, RTP_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_TP_CTRL,
+					      PRBS31R_EN, 0);
+				XPCS_RGWR_VAL(pdata, PCS_KR_CTRL, PRBS9RXEN, 1);
+			} else if (strncasecmp(token, "RANDOM",
+				min_t(size_t, len, 6UL)) == 0) {
+				pr_info("random not supported");
+			}
+		}
+		break;
+	case 11:
+		serdes_rx_test_pattern_cfg_get(pdata);
+		break;
+	case 12:
+		serdes_rx_bit_error_get(pdata);
+		break;
+	default:
+		bert_help();
+		break;
+	}
+
+	return count;
+}
+
+static DEVICE_ATTR_RW(bert);
+
 static struct attribute *xpcs_attrs[] = {
 	&dev_attr_xpcs_cl73_info.attr,
 	&dev_attr_xpcs_status.attr,
@@ -1135,6 +1402,7 @@ static struct attribute *xpcs_attrs[] = {
 	&dev_attr_txeq.attr,
 	&dev_attr_mode.attr,
 	&dev_attr_cl37.attr,
+	&dev_attr_bert.attr,
 	NULL,
 };
 
